rfcs/20220925-bounded-dynamism.md

```diff
@@ -0,0 +1,300 @@
+all the relevant ops.
+
+## (P3) get\_dimension\_size / set\_dimension\_size ops
+
+The `get_dimension_size` op takes a tensor and a dimension index and returns the
+actual size of the dimension at runtime as an `i32` type scalar.
+
+The following example returns the size of the result after concatenating input
+that has up to `16` elements with self and returns the actual runtime size of
+the concatenation result.
+
+```
+func.func @self_concat_size(%data: tensor<?xi32, #stablehlo.type_extensions<bounds = [16]>>) -> tensor<i32> {
+  %concat = "stablehlo.concatenate"(%data, %data) {dimension = 0 : i64}
+    : (tensor<?xi32, #stablehlo.type_extensions<bounds = [16]>>,
+       tensor<?xi32, #stablehlo.type_extensions<bounds = [16]>>)
+      -> tensor<?xi32, #stablehlo.type_extensions<bounds = [32]>>
+
+  %result = stablehlo.get_dimension_size %concat, dim = 0
+    : (tensor<?xi32, #stablehlo.type_extensions<bounds = [32]>>) -> tensor<i32>
+
+  func.return %result : tensor<i32>
+}
+```
+
+The `set_dimension_size` op takes a bounded tensor, runtime size and a dimension
+index and returns a tensor whose logical size of the particular dimension is set
+to the specified size. This size needs to be less than or equal to the static
+size or bound for the dimension. This operation can be thought as either a slice
+or pad operation depending on if the earlier logical dimension size is larger or
+smaller, respectively. In case the dimension size is increased, the padded
+values are undefined.
+
+In the following example, `set_dimension_size` op is used to set the logical
+size of the first dimension so that it performs a sum reduction on the first
+`batch_size` elements in the input. With data argument `[1, 2, 3, 4]` and
+batch\_size argument `2`, the following function returns `3` but it returns `6`
+for the same data argument when the batch\_size is `3`. The `set_dimension_size`
+op also sets the bound on the returned tensor. This bound depends on operand's
+static size if the operand shape is static. It is `4` in this example. If the
+operand dimension is not static, then the returned tensor has same type as the
+operand.
+
+```
+func.func @dynamic_sum(%data: tensor<4xi32>, %batch_size: tensor<i32>) -> tensor<i32> {
+  %dynamic_data = stablehlo.set_dimension_size %data, %batch_size, dim = 0
+    : (tensor<4xi32>, tensor<i32>) -> tensor<?xi32, #stablehlo.type_extensions<bounds = [4]>>
+
+  %zero = stablehlo.constant dense<0> : tensor<i32>
+  %sum = "stablehlo.reduce"(%dynamic_data, %zero) ({
+   ^bb0(%arg1: tensor<i32>, %arg2: tensor<i32>):
+    %add = stablehlo.add %arg1, %arg2 : tensor<i32>
+    "stablehlo.return"(%add) : (tensor<i32>) -> ()
+  }) {dimensions = dense<[0]> : tensor<1xi64>}
+    : (tensor<?xi32, #stablehlo.type_extensions<bounds = [4]>>,
+       tensor<i32>) -> tensor<i32>
+  func.return %sum : tensor<i32>
+}
+```
+
+## (P4) Prefer generic dynamic ops over set\_dimension\_size op
```

Comment:
To echo @burmako: this feels a bit out of place from the rest of the discussion. 

Consider titling this: "Comparison with `real_dynamic_slice` and `dynamic_pad`". That would let you start this section like: "In addition to set_dimension_size" and "get_dimension_size", StableHLO producers may also use `real_dynamic_slice` and `dynamic_pad` to perform operations on dynamically shaped tensors. For example, the above `@dynamic_sum` computation can ...

Are there any negatives to using the dynamic ops? If not, why not just adopt them directly and skip the old ops?

Reply:
This is briefly covered in the last section.

Main motivation is to allow faster adoption of StableHLO and gradually move to the desired state. Given that these are only two ops, they don't have a significant maintenance overhead.